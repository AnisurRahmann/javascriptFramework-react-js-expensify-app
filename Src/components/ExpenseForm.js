import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';

import 'react-dates/initialize';

 export default class ExpenseForm extends React.Component {
     constructor(props){
     super(props)
     this.state= {
        descriptions:props.expenses?props.expenses.description:'',
        amount:props.expenses?(props.expenses.amount/100).toString():'',
        note:props.expenses?props.expenses.note:'',
        createdAt:props.expenses?moment(props.expenses.createdAt):moment(),
        calenderFocused:false,
        error:''
    }

     }
     
     onDescriptionChange =  (e) => {
        const descriptions=e.target.value;
        this.setState(()=>({descriptions}))
     }
     onAmountChange = (e) => {
         const amount = e.target.value
         if(!amount||amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(() => ({amount}));
         }

     }
     onNoteChange = (e) => {
         const note=e.target.value;
         this.setState(()=>({note}));
     }
     onDatechange = (createdAt) => {
         if(createdAt){
            this.setState(()=>({createdAt}));
         }
         
     }
     onFocusChange = ({focused}) => {
         this.setState (()=>({calenderFocused:focused}))
     }
     onSubmit = (e) => {
         e.preventDefault();
        if(!this.state.descriptions || !this.state.amount){
            this.setState(() => ({error:'please fill up requiremnts'}))
        }else{
            this.setState(() => ({error:''}));
            this.props.onSubmit ({
                description:this.state.descriptions,
                amount:parseFloat(this.state.amount) * 10,
                createdAt:this.state.createdAt.valueOf(),
                note:this.state.note,


            })
        }
     }
     render(){
         return(
         <div>
         {this.state.error && <p>{this.state.error}</p>}
             <form onSubmit={this.onSubmit}>
              <input type="text"
                placeholder="descriptions"
                value={this.state.descriptions}
                autoFocus
                onChange={this.onDescriptionChange}
                />
              <input
                type="number"
                placeholder="amount"
                value={this.state.amount}
                onChange={this.onAmountChange}
                />
               <SingleDatePicker
                 date={this.state.createdAt}
                 onDateChange={this.onDatechange}
                 focused={this.state.calenderFocused}
                 onFocusChange={this.onFocusChange}
                 numberOfMonths={1}
                 isOutsideRange={()=>false}
               
               />
              <textarea
                placeholder="add a note (optional)"
                onChange={this.onNoteChange}></textarea>
              <button>Add</button>
             </form>
         </div>)
     }
 }