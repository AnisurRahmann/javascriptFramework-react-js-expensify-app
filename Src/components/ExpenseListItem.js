import React from 'react';
import ReactDOM from 'react-dom';
import {connect} from 'react-redux';

import {Link} from 'react-router-dom';

export const ExpenseListItem = ({description,id,amount,createdAt,}) => {
    return (
       <div>
       <Link to={`edite/${id}`}>
          <h3>{description}</h3>
       </Link>
        
            <p>{amount}-{createdAt}</p>
           
       </div>
    )
}
export default connect()(ExpenseListItem );