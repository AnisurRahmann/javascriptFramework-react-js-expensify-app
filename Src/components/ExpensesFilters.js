import React from 'react';
import {connect} from 'react-redux';
import  {DateRangePicker} from 'react-dates';
import uuid from 'uuid';
import {setTextFilter,sortByAmount,sortByDate,setStartDate,setEndDate} from '../actions/filters';

export class ExpensesFilters extends React.Component{
    state={
        calenderFocused:null
    }
    onDatesChange = ({startDate,endDate}) =>{
        this.props.setStartDate(startDate);
        console.log(this.props.startDate)
        this.props.setEndDate(endDate);
    }
    onFocusChange = (calenderFocused) => {
        this.setState(()=>({calenderFocused}))
    }
    onTextFilterChange =(e) => {
        this.props.setTextFilter(e.target.value);
    }
    onSortChange=(e)=>{
        if(e.target.value === 'Amount'){
        this.props.sortByAmount();
        } else if(e.target.value === 'date'){
            this.props.sortByDate();
        }
    }



    render(){
        return(
            <div>
                    <input type="text"
                    value={this.props.filters.text}
                    onChange={this.onTextFilterChange }
                />
                <select value={this.props.filters.sortBy}
                    onChange={this.onSortChange}>
                    <option value="Amount">Amount</option>
                    <option value="date">Date</option>
                </select>
                <DateRangePicker  
                  startDate={this.props.filters.startDate}
                  endDate={this.props.filters.endDate}
                  onDatesChange={this.onDatesChange}
                  focusedInput={this.state.calenderFocused}
                  onFocusChange={this.onFocusChange}
                  isOutsideRange={()=>false}
                  numberOfMonths={1}
                  showClearDates={true}
                 
                />
        </div>
        )
    }
}

const mapStateToProps =  (state) => {
    return {
        filters:state.filters
    };
}
const mapDispatchToProps = (dispatch) => {
    return {
        setStartDate:(starDate)=>dispatch(setStartDate(starDate)),
        setEndDate:(endDate)=>dispatch(setEndDate(endDate)),
        setTextFilter:(text)=>dispatch(setTextFilter(text)),
        sortByAmount:()=>dispatch(sortByAmount()),
        sortByDate:()=>dispatch(sortByDate())

        
    }
}
export default connect(mapStateToProps,mapDispatchToProps)(ExpensesFilters);