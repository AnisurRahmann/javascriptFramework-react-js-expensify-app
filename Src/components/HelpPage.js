import React from 'react';

const HelpPage = () => (
    <div>
        this is from Help Page
    </div>
);
export default HelpPage;