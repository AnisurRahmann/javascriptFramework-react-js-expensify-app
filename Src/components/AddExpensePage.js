import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {addExpense} from '../actions/expenses'

export  class AddExpensePage extends React.Component{
    onSubmit=(expenses)=>{
        this.props.addExpense(expenses)
            this.props.history.push('/');
        }
    render(){
        return(
            <div>
       <ExpenseForm 
         onSubmit={this.onSubmit} />
            </div>
        )
    }



  
}


const mapDispatchToProps = (dispatch) =>{
  return{
      addExpense:(expenses)=>dispatch(addExpense(expenses))
  }
}
export default connect(undefined,mapDispatchToProps)(AddExpensePage);