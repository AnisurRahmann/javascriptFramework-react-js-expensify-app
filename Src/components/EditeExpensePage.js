import React from 'react';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import {editeExpense,removeExpense} from '../actions/expenses';

export class EditeExpensePage extends React.Component{
    onSubmit=(expenses)=> {
        this.props.editeExpense(this.props.expenses.id,expenses);
        this.props.history.push('/')
    }
    onRemove = () =>  {
        this.props.removeExpense({id:this.props.expenses.id});
        this.props.history.push('/');
    }
    render(){
        return(
            <div>
     <ExpenseForm
        expenses={this.props.expenses}
        onSubmit={this.onSubmit}
      />
      <button onClick={this.onRemove }>DELETE</button>
    </div>
        )
    }


}

const mapStateToProps = (state,props) => {
   return {
       expenses:state.expenses.find((expnese)=>expnese.id===props.match.params.id)
   }
}
const mapDispatchToProps = (dispatch,props) => {
  return {
      editeExpense:(expenses)=>dispatch(editeExpense(id,expenses)),
      removeExpense:(data)=>dispatch(removeExpense(data))
  }
}
export default connect(mapStateToProps,mapDispatchToProps)(EditeExpensePage);