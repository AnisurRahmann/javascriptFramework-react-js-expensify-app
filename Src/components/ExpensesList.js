import React from 'react';
import ReactDOM from 'react-dom';
import {connect} from  'react-redux';
import ExpenseListItem from './ExpenseListItem';
import SelectExpenses from '../selectors/expenses' ;
export const ExpensesList = (props) => {
    return (
        <div>
        {
            props.expenses.length === 0 ? (
                <h1>no Expenses List</h1>
            ) : (
                    props.expenses.map((expense) => {
                        return  <ExpenseListItem key={expense.id}{...expense}/>
                        })
                )
        }
        </div>
    )
}
const  mapStateToProps =  (state) => {
    return {
        expenses:SelectExpenses(state.expenses,state.filters)
    }
}

export default connect(mapStateToProps)(ExpensesList);