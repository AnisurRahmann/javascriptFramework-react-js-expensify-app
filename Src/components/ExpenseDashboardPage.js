import React from 'react';
import ExpensesList from './ExpensesList';
import ExpensesFilters from './ExpensesFilters';

const ExpenseDashboardPage = () => (
    <div>
       <ExpensesFilters />
       <ExpensesList />
    </div>
);
export default ExpenseDashboardPage;