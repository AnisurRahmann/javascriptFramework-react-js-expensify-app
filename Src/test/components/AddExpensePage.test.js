import {shallow} from 'enzyme';
import {AddExpensePage} from '../../components/AddExpensePage';
import React from 'react';
import expenses from '../reducer/fixtures/expesnses';

let addExpense ,history,wrapper;
beforeEach(()=>{
    addExpense=jest.fn();
    history={push:jest.fn()};
    wrapper=shallow( <AddExpensePage  addExpense={addExpense} history={history} /> )
})
test('test add expenses',() => {
  
    expect(wrapper).toMatchSnapshot();
})
test('check onSubmit',() => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1])
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(addExpense).toHaveBeenLastCalledWith(expenses[1])
})