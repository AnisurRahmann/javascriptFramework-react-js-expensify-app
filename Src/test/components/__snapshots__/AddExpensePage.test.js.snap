// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test add expenses 1`] = `
<div>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
